src/memory_pool.c

```diff
@@ -7,43 +7,99 @@
  *
  */
 
-#include "libumf.h"
-#include "memory_pool_internal.h"
-#include "utils_assert.h"
-
+#include <umf/base.h>
 #include <umf/memory_pool.h>
 #include <umf/memory_pool_ops.h>
 
 #include <assert.h>
-#include <stdlib.h>
+#include <string.h>
 
 #include "base_alloc_global.h"
+#include "ctl/ctl.h"
+#include "libumf.h"
 #include "memory_pool_internal.h"
 #include "memory_provider_internal.h"
 #include "provider_tracking.h"
+#include "utils_assert.h"
+#include "utils_concurrency.h"
+#include "utils_log.h"
+
+#define UMF_DEFAULT_SIZE 100
+#define UMF_DEFAULT_LEN 100
+
+utils_mutex_t ctl_mtx;
+static UTIL_ONCE_FLAG mem_pool_ctl_initialized = UTIL_ONCE_FLAG_INIT;
+
+char CTL_DEFAULT_ENTRIES[UMF_DEFAULT_SIZE][UMF_DEFAULT_LEN] = {0};
+char CTL_DEFAULT_VALUES[UMF_DEFAULT_SIZE][UMF_DEFAULT_LEN] = {0};
+
+void ctl_init(void) { utils_mutex_init(&ctl_mtx); }
 
 static int CTL_SUBTREE_HANDLER(by_handle_pool)(void *ctx,
                                                umf_ctl_query_source_t source,
-                                               void *arg,
+                                               void *arg, size_t size,
                                                umf_ctl_index_utlist_t *indexes,
                                                const char *extra_name,
                                                umf_ctl_query_type_t queryType) {
     (void)indexes, (void)source;
     umf_memory_pool_handle_t hPool = (umf_memory_pool_handle_t)ctx;
-    hPool->ops.ctl(hPool, /*unused*/ 0, extra_name, arg, queryType);
+    hPool->ops.ctl(hPool->pool_priv, /*unused*/ 0, extra_name, arg, size,
+                   queryType);
+    return 0;
+}
+
+static int CTL_SUBTREE_HANDLER(default)(void *ctx,
+                                        umf_ctl_query_source_t source,
+                                        void *arg, size_t size,
+                                        umf_ctl_index_utlist_t *indexes,
+                                        const char *extra_name,
+                                        umf_ctl_query_type_t queryType) {
+    (void)indexes, (void)source, (void)ctx;
+    utils_init_once(&mem_pool_ctl_initialized, ctl_init);
+    utils_mutex_lock(&ctl_mtx);
+    if (queryType == CTL_QUERY_WRITE) {
+        int i = 0;
+        for (; i < UMF_DEFAULT_SIZE; i++) {
+            if (CTL_DEFAULT_ENTRIES[i][0] == '\0' ||
+                strcmp(CTL_DEFAULT_ENTRIES[i], extra_name) == 0) {
+                strncpy(CTL_DEFAULT_ENTRIES[i], extra_name, UMF_DEFAULT_LEN);
+                strncpy(CTL_DEFAULT_VALUES[i], arg, UMF_DEFAULT_LEN);
+                break;
+            }
+        }
+        if (UMF_DEFAULT_SIZE == i) {
+            LOG_ERR("Default entries array is full");
+            return -1;
```

Comment:
please decide if this function returns int error codes, or umf result

Reply:
Done.